### ISSUE TITLE
Why The padding side must be set to "left" during training

### ISSUE THREAD

**Issue by lujiazho** [om-ai-lab/VLM-R1#290]
I saw in the grpo_trainer python file

![Image](https://github.com/user-attachments/assets/750a51b4-97ff-46d1-bc11-e2250cb7b580)

![Image](https://github.com/user-attachments/assets/51731fe3-8e1f-4811-99d6-5a4762af15cc)

But why? I thought the standard practice is to use **right padding for training** and **left padding for inference** to avoid problems. This is also the way how MedGemma is fine-tuned: https://comfyai.app/article/llm-questions-and-answers/left-padding-or-right-padding

**Comment by wyddmw** [comment]
> I saw in the grpo_trainer python file
> 
> ![Image](https://github.com/user-attachments/assets/750a51b4-97ff-46d1-bc11-e2250cb7b580)
> 
> ![Image](https://github.com/user-attachments/assets/51731fe3-8e1f-4811-99d6-5a4762af15cc)
> 
> But why? I thought the standard practice is to use **right padding for training** and **left padding for inference** to avoid problems. This is also the way how MedGemma is fine-tuned: https://comfyai.app/article/llm-questions-and-answers/left-padding-or-right-padding

I think this is because the stage of training we are discussing here actually does the inference to generate the response where the padding_side is set to left.

**Comment by lujiazho** [comment]
Thanks for your reply. I think I figured out why:
1. In SFT, it's teacher-forcing for calculating the loss, and the training is executed in a parallel way by inputting all prompts + ground truth outputs into the model, which makes it the **right padding** during training.
2. But in GRPO, it's reward-based (either rule-based or LLM-as-a-reward), so even in the training phase, we still sample the outputs just like doing inference, which makes it **left padding**.

One more thing about the inference of the reward model is that (in around 647 lines of grpo_trainer.py)
```
reward_inputs = reward_processing_class(
    texts, return_tensors="pt", padding=True, padding_side="right", add_special_tokens=False
)
```
It's the right padding because the LLM-based reward model is just to give a single reward score (with ORM), not generating a sentence, and its training is based on the right padding, which makes it still the **right padding during inference**.
